pkg/operator/ceph/pool/radosnamespace/controller.go

```diff
@@ -0,0 +1,312 @@
+
+	// DELETE: the CR was deleted
+	if !cephBlockPoolRadosNamespace.GetDeletionTimestamp().IsZero() {
+		logger.Debugf("delete ceph blockpool rados namespace %q", cephBlockPoolRadosNamespace.Name)
+		// On external cluster, we don't delete the rados namespace, it has to be deleted manually
+		if cephCluster.Spec.External.Enable {
+			logger.Warning("external rados namespace deletion is not supported, delete it manually")
+		} else {
+			err := r.deleteRadosNamespace(cephBlockPoolRadosNamespace)
+			if err != nil {
+				if strings.Contains(err.Error(), opcontroller.UninitializedCephConfigError) {
+					logger.Info(opcontroller.OperatorNotInitializedMessage)
+					return opcontroller.WaitForRequeueIfOperatorNotInitialized, nil
+				}
+				return reconcile.Result{}, errors.Wrapf(err, "failed to delete ceph blockpool rados namespace %q", cephBlockPoolRadosNamespace.Name)
+			}
+		}
+		err = csi.SaveClusterConfig(r.context.Clientset, buildClusterID(cephBlockPoolRadosNamespace), r.clusterInfo, nil)
+		if err != nil {
+			return reconcile.Result{}, errors.Wrap(err, "failed to save cluster config")
+		}
+		// Remove finalizer
+		err = opcontroller.RemoveFinalizer(r.opManagerContext, r.client, cephBlockPoolRadosNamespace)
+		if err != nil {
+			return reconcile.Result{}, errors.Wrap(err, "failed to remove finalizer")
+		}
+
+		// Return and do not requeue. Successful deletion.
+		return reconcile.Result{}, nil
+	}
+
+	if !cephCluster.Spec.External.Enable {
+		// Build the NamespacedName to fetch the CephBlockPool and make sure it exists, if not we cannot
+		// create the rados namespace
+		cephBlockPool := &cephv1.CephBlockPool{}
+		cephBlockPoolRadosNamespacedName := types.NamespacedName{Name: cephBlockPoolRadosNamespace.Spec.BlockPoolName, Namespace: request.Namespace}
+		err = r.client.Get(r.opManagerContext, cephBlockPoolRadosNamespacedName, cephBlockPool)
+		if err != nil {
+			if kerrors.IsNotFound(err) {
+				return reconcile.Result{}, errors.Wrapf(err, "failed to fetch ceph blockpool %q, cannot create rados namespace %q", cephBlockPoolRadosNamespace.Spec.BlockPoolName, cephBlockPoolRadosNamespace.Name)
+			}
+			// Error reading the object - requeue the request.
+			return reconcile.Result{}, errors.Wrap(err, "failed to get cephBlockPoolRadosNamespace")
+		}
+
+		// If the cephBlockPool is not ready to accept commands, we should wait for it to be ready
+		if cephBlockPool.Status.Phase != cephv1.ConditionReady {
+			// We know the CR is present so it should a matter of second for it to become ready
+			return reconcile.Result{Requeue: true, RequeueAfter: 10 * time.Second}, errors.Wrapf(err, "failed to fetch ceph blockpool %q, cannot create rados namespace %q", cephBlockPoolRadosNamespace.Spec.BlockPoolName, cephBlockPoolRadosNamespace.Name)
+		}
+	}
+	if cephCluster.Spec.External.Enable {
+		logger.Debug("external rados namespace creation is not supported, create it manually, the controller will assume it's there")
```

Comment:
The rook operator can operate on multiple resources in multiple namespaces. All info/debug messages must list the namespaced name so context is clear.